divide by zero error when compiling to verilog/vhdl

When compiling 'TopEntity.hs' (see attachment) to verilog or vhdl, clash exits with the following error:
```
<no location info>: error:
    Other error:
    divide by zero
```
We can run and sample topEntity with clashi, so it passes the typecheck.
We tried this with commit fff460, built with stack, and clash 0.99.3, built with cabal.
[Divide_by_zero_error.zip](https://github.com/clash-lang/clash-compiler/files/3053379/Divide_by_zero_error.zip)

You have this `div` in `setRegister`, let's add a type wildcard and see if clash will tell us what it's dividing here:
```haskell
setRegister modtype num (RECIEVED (_,dat)) = SET (getReg(modtype, num `div` (2::_)), num `mod` 2, dat)
```
```
TopEntity.hs:30:81: error:
    • Found type wildcard ‘_’ standing for ‘Unsigned 1’
```
Mmm a 1 bit wide 2...
```
> 2 :: Unsigned 1
0
```
That's probably not what you meant to write.
And clash is crashing on it because it tries to evaluate all constants at compile time.

I would still consider this a bug in clash, because it shouldn't crash on something like this.
Or at least not with such an unhelpful error.

--- a/clash-ghc/src-ghc/Clash/GHC/Evaluator.hs
+++ b/clash-ghc/src-ghc/Clash/GHC/Evaluator.hs
@@ -22,6 +22,8 @@ module Clash.GHC.Evaluator where
 
 import           Control.Applicative (liftA2)
 import           Control.Concurrent.Supply  (Supply,freshId)
+import           Control.DeepSeq            (force)
+import           Control.Exception          (ArithException(..), Exception, tryJust, evaluate)
 import           Control.Monad              (ap)
 import           Control.Monad.Trans.Except (runExcept)
 import           Data.Bits
@@ -69,6 +71,7 @@ import           Clash.Core.Evaluator
 import           Clash.Core.Literal  (Literal (..))
 import           Clash.Core.Name
   (Name (..), NameSort (..), mkUnsafeSystemName)
+import           Clash.Core.Pretty   (showPpr)
 import           Clash.Core.Term
   (Pat (..), PrimInfo (..), Term (..), WorkInfo (..))
 import           Clash.Core.Type
@@ -150,16 +153,17 @@ reduceConstant isSubj tcm h k nm pInfo tys args = case nm of
        in  reduce (integerToIntLiteral (toInteger $ I# c))
 
   "GHC.Prim.quotInt#" | Just (i,j) <- intLiterals args
-    -> reduce (integerToIntLiteral (i `quot` j))
+    -> reduce $ catchDivByZero (integerToIntLiteral (i `quot` j))
   "GHC.Prim.remInt#" | Just (i,j) <- intLiterals args
-    -> reduce (integerToIntLiteral (i `rem` j))
+    -> reduce $ catchDivByZero (integerToIntLiteral (i `rem` j))
   "GHC.Prim.quotRemInt#" | Just (i,j) <- intLiterals args
     -> let (_,tyView -> TyConApp tupTcNm tyArgs) = splitFunForallTy ty
            (Just tupTc) = lookupUniqMap tupTcNm tcm
            [tupDc] = tyConDataCons tupTc
            (q,r)   = quotRem i j
            ret     = mkApps (Data tupDc) (map Right tyArgs ++
-                    [Left (integerToIntLiteral q), Left (integerToIntLiteral r)])
+                    [Left $ catchDivByZero (integerToIntLiteral q)
+                    ,Left $ catchDivByZero (integerToIntLiteral r)])
        in  reduce ret
 
   "GHC.Prim.andI#" | Just (i,j) <- intLiterals args
@@ -297,16 +301,17 @@ reduceConstant isSubj tcm h k nm pInfo tys args = case nm of
                    , Left (Literal . WordLiteral . toInteger $ W# l)])
 
   "GHC.Prim.quotWord#" | Just (i,j) <- wordLiterals args
-    -> reduce (integerToWordLiteral (i `quot` j))
+    -> reduce $ catchDivByZero (integerToWordLiteral (i `quot` j))
   "GHC.Prim.remWord#" | Just (i,j) <- wordLiterals args
-    -> reduce (integerToWordLiteral (i `rem` j))
+    -> reduce $ catchDivByZero (integerToWordLiteral (i `rem` j))
   "GHC.Prim.quotRemWord#" | Just (i,j) <- wordLiterals args
     -> let (_,tyView -> TyConApp tupTcNm tyArgs) = splitFunForallTy ty
            (Just tupTc) = lookupUniqMap tupTcNm tcm
            [tupDc] = tyConDataCons tupTc
            (q,r)   = quotRem i j
            ret     = mkApps (Data tupDc) (map Right tyArgs ++
-                    [Left (integerToWordLiteral q), Left (integerToWordLiteral r)])
+                    [Left $ catchDivByZero (integerToWordLiteral q)
+                    ,Left $ catchDivByZero (integerToWordLiteral r)])
        in  reduce ret
   "GHC.Prim.quotRemWord2#" | [i,j,k'] <- wordLiterals' args
     -> let (_,tyView -> TyConApp tupTcNm tyArgs) = splitFunForallTy ty
@@ -318,8 +323,8 @@ reduceConstant isSubj tcm h k nm pInfo tys args = case nm of
            !(# x, y #) = quotRemWord2# a b c
        in  reduce $
            mkApps (Data tupDc) (map Right tyArgs ++
-                   [ Left (Literal . WordLiteral . toInteger $ W# x)
-                   , Left (Literal . WordLiteral . toInteger $ W# y)])
+                   [ Left $ catchDivByZero (Literal . WordLiteral . toInteger $ W# x)
+                   , Left $ catchDivByZero (Literal . WordLiteral . toInteger $ W# y)])
 
   "GHC.Prim.and#" | Just (i,j) <- wordLiterals args
     -> reduce (integerToWordLiteral (i .&. j))
@@ -884,8 +889,8 @@ reduceConstant isSubj tcm h k nm pInfo tys args = case nm of
            (q,r) = quotRem i j
     in reduce $
          mkApps (Data tupDc) (map Right tyArgs ++
-                [ Left (integerToIntegerLiteral q)
-                , Left (integerToIntegerLiteral r)])
+                [ Left $ catchDivByZero (integerToIntegerLiteral q)
+                , Left $ catchDivByZero (integerToIntegerLiteral r)])
 
   "GHC.Integer.Type.plusInteger" | Just (i,j) <- integerLiterals args
     -> reduce (integerToIntegerLiteral (i+j))
@@ -901,16 +906,16 @@ reduceConstant isSubj tcm h k nm pInfo tys args = case nm of
     -> reduce (integerToIntegerLiteral (negate i))
 
   "GHC.Integer.Type.divInteger" | Just (i,j) <- integerLiterals args
-    -> reduce (integerToIntegerLiteral (i `div` j))
+    -> reduce $ catchDivByZero (integerToIntegerLiteral (i `div` j))
 
   "GHC.Integer.Type.modInteger" | Just (i,j) <- integerLiterals args
-    -> reduce (integerToIntegerLiteral (i `mod` j))
+    -> reduce $ catchDivByZero (integerToIntegerLiteral (i `mod` j))
 
   "GHC.Integer.Type.quotInteger" | Just (i,j) <- integerLiterals args
-    -> reduce (integerToIntegerLiteral (i `quot` j))
+    -> reduce $ catchDivByZero (integerToIntegerLiteral (i `quot` j))
 
   "GHC.Integer.Type.remInteger" | Just (i,j) <- integerLiterals args
-    -> reduce (integerToIntegerLiteral (i `rem` j))
+    -> reduce $ catchDivByZero (integerToIntegerLiteral (i `rem` j))
 
   "GHC.Integer.Type.divModInteger" | Just (i,j) <- integerLiterals args
     -> let (_,tyView -> TyConApp ubTupTcNm [liftedKi,_,intTy,_]) = splitFunForallTy ty
@@ -920,8 +925,8 @@ reduceConstant isSubj tcm h k nm pInfo tys args = case nm of
        in  reduce $
            mkApps (Data ubTupDc) [ Right liftedKi, Right liftedKi
                                  , Right intTy,    Right intTy
-                                 , Left (Literal (IntegerLiteral d))
-                                 , Left (Literal (IntegerLiteral m))
+                                 , Left $ catchDivByZero (Literal (IntegerLiteral d))
+                                 , Left $ catchDivByZero (Literal (IntegerLiteral m))
                                  ]
 
   "GHC.Integer.Type.gtInteger" | Just (i,j) <- integerLiterals args
@@ -1628,11 +1633,11 @@ reduceConstant isSubj tcm h k nm pInfo tys args = case nm of
   "Clash.Sized.Internal.BitVector.quot#"
     | Just (_, kn) <- extractKnownNat tcm tys
     , Just val <- reifyNat kn (liftBitVector2 (BitVector.quot#) ty tcm tys args)
-    -> reduce val
+    -> reduce $ catchDivByZero val
   "Clash.Sized.Internal.BitVector.rem#"
     | Just (_, kn) <- extractKnownNat tcm tys
     , Just val <- reifyNat kn (liftBitVector2 (BitVector.rem#) ty tcm tys args)
-    -> reduce val
+    -> reduce $ catchDivByZero val
   "Clash.Sized.Internal.BitVector.toInteger#"
     | Just (_, kn) <- extractKnownNat tcm tys
     , [i] <- bitVectorLiterals' args
@@ -1788,11 +1793,11 @@ reduceConstant isSubj tcm h k nm pInfo tys args = case nm of
   "Clash.Sized.Internal.Index.quot#"
     | Just (nTy,kn) <- extractKnownNat tcm tys
     , Just (i,j) <- indexLiterals args
-    -> reduce (mkIndexLit ty nTy kn (i `quot` j))
+    -> reduce $ catchDivByZero (mkIndexLit ty nTy kn (i `quot` j))
   "Clash.Sized.Internal.Index.rem#"
     | Just (nTy,kn) <- extractKnownNat tcm tys
     , Just (i,j) <- indexLiterals args
-    -> reduce (mkIndexLit ty nTy kn (i `rem` j))
+    -> reduce $ catchDivByZero (mkIndexLit ty nTy kn (i `rem` j))
   "Clash.Sized.Internal.Index.toInteger#"
     | [PrimVal nm' _ _ [_, Lit (IntegerLiteral i)]] <- args
     , nm' == "Clash.Sized.Internal.Index.fromInteger#"
@@ -1917,19 +1922,19 @@ reduceConstant isSubj tcm h k nm pInfo tys args = case nm of
   "Clash.Sized.Internal.Signed.quot#"
     | Just (_, kn) <- extractKnownNat tcm tys
     , Just val <- reifyNat kn (liftSigned2 (Signed.quot#) ty tcm tys args)
-    -> reduce val
+    -> reduce $ catchDivByZero val
   "Clash.Sized.Internal.Signed.rem#"
     | Just (_, kn) <- extractKnownNat tcm tys
     , Just val <- reifyNat kn (liftSigned2 (Signed.rem#) ty tcm tys args)
-    -> reduce val
+    -> reduce $ catchDivByZero val
   "Clash.Sized.Internal.Signed.div#"
     | Just (_, kn) <- extractKnownNat tcm tys
     , Just val <- reifyNat kn (liftSigned2 (Signed.div#) ty tcm tys args)
-    -> reduce val
+    -> reduce $ catchDivByZero val
   "Clash.Sized.Internal.Signed.mod#"
     | Just (_, kn) <- extractKnownNat tcm tys
     , Just val <- reifyNat kn (liftSigned2 (Signed.mod#) ty tcm tys args)
-    -> reduce val
+    -> reduce $ catchDivByZero val
   "Clash.Sized.Internal.Signed.toInteger#"
     | [PrimVal nm' _ _ [_, Lit (IntegerLiteral i)]] <- args
     , nm' == "Clash.Sized.Internal.Signed.fromInteger#"
@@ -2116,11 +2121,11 @@ reduceConstant isSubj tcm h k nm pInfo tys args = case nm of
   "Clash.Sized.Internal.Unsigned.quot#"
     | Just (_, kn) <- extractKnownNat tcm tys
     , Just val <- reifyNat kn (liftUnsigned2 (Unsigned.quot#) ty tcm tys args)
-    -> reduce val
+    -> reduce $ catchDivByZero val
   "Clash.Sized.Internal.Unsigned.rem#"
     | Just (_, kn) <- extractKnownNat tcm tys
     , Just val <- reifyNat kn (liftUnsigned2 (Unsigned.rem#) ty tcm tys args)
-    -> reduce val
+    -> reduce $ catchDivByZero val
   "Clash.Sized.Internal.Unsigned.toInteger#"
     | [PrimVal nm' _ _ [_, Lit (IntegerLiteral i)]] <- args
     , nm' == "Clash.Sized.Internal.Unsigned.fromInteger#"
@@ -3271,6 +3276,21 @@ reduceConstant isSubj tcm h k nm pInfo tys args = case nm of
                    in  Just (h2,k,e')
     reduceWHNF' h' e = let (h2,[],e') = whnf reduceConstant tcm isSubj (h',[],e)
                        in  Just (h2,k,e')
+
+    makeUndefinedIf :: Exception e => (e -> Bool) -> Term -> Term
+    makeUndefinedIf wantToHandle tm =
+      case unsafeDupablePerformIO $ tryJust selectException (evaluate $ force tm) of
+        Right b -> b
+        Left e -> trace (msg e) (undefinedTm resTy)
+      where
+        resTy = getResultTy tcm ty tys
+        selectException e | wantToHandle e = Just e
+                          | otherwise = Nothing
+        msg e = unlines ["Warning: caught exception: \"" ++ show e ++ "\" while trying to evaluate: "
+                        , showPpr (mkApps (Prim nm pInfo) (map (Left . valToTerm) args))
+                        ]
+
+    catchDivByZero = makeUndefinedIf (==DivideByZero)
 
 typedLiterals' :: (Value -> Maybe a) -> [Value] -> [a]
 typedLiterals' typedLiteral = mapMaybe typedLiteral

--- a/clash-lib/prims/common/Clash_Transformations.json
+++ b/clash-lib/prims/common/Clash_Transformations.json
@@ -6,4 +6,12 @@
     , "template"  : "~ERRORO"
     }
   }
+, { "BlackBox" :
+    { "name"      : "Clash.Transformations.undefined"
+    , "workInfo"  : "Constant"
+    , "kind"      : "Expression"
+    , "type"      : "undefined :: forall a . a"
+    , "template"  : "~ERRORO"
+    }
+  }
 ]

--- a/clash-lib/src/Clash/Core/Type.hs
+++ b/clash-lib/src/Clash/Core/Type.hs
@@ -515,7 +515,7 @@ litView _ (LitTy (NumTy i))                = Just i
 litView m (reduceTypeFamily m -> Just ty') = litView m ty'
 litView _ _ = Nothing
 
--- | The type of GHC.Err.undefined :: forall a . a
+-- | The type @forall a . a@
 undefinedTy ::Type
 undefinedTy =
   let aNm = mkUnsafeSystemName "a" 0

--- a/clash-lib/src/Clash/Core/Util.hs
+++ b/clash-lib/src/Clash/Core/Util.hs
@@ -907,7 +907,7 @@ primCo ty = Prim "_CO_" (PrimInfo ty WorkNever)
 undefinedTm
   :: Type
   -> Term
-undefinedTm = TyApp (Prim "Clash.Transformation." (PrimInfo undefinedTy WorkNever))
+undefinedTm = TyApp (Prim "Clash.Transformations.undefined" (PrimInfo undefinedTy WorkNever))
 
 substArgTys
   :: DataCon

--- a/tests/shouldwork/Numbers/DivideByZero.hs
+++ b/tests/shouldwork/Numbers/DivideByZero.hs
@@ -1,0 +1,35 @@
+-- Test that the compiler doesn't crash on any DivideByZero exceptions
+module DivideByZero where
+
+import Clash.Prelude
+import Clash.Explicit.Testbench
+import Data.Word
+import Data.Int
+
+topEntity = pure @(Signal System)
+  (test @(Unsigned 7)
+  ,test @(Signed 9)
+  ,test @(BitVector 11)
+  ,test @(Index 13)
+  ,test @Int
+  ,test @Integer
+  ,test @Word16
+  ,test @Int32
+  )
+{-# NOINLINE topEntity #-}
+
+test :: Integral a => (a,a,a,a)
+test = (div0,quot0,mod0,rem0)
+  where
+    div0  = 1 `div`  0
+    quot0 = 2 `quot` 0
+    mod0  = 3 `mod`  0
+    rem0  = 4 `rem`  0
+
+testBench :: Signal System Bool
+testBench = done
+  where
+    expectedOutput = outputVerifier clk rst $ (undefined,undefined,undefined,undefined,undefined,undefined,undefined,undefined) :> Nil
+    done           = expectedOutput topEntity
+    clk            = tbSystemClockGen (not <$> done)
+    rst            = systemResetGen

--- a/testsuite/Main.hs
+++ b/testsuite/Main.hs
@@ -193,6 +193,7 @@ runClashTest =
       , clashTestGroup "Numbers"
         [ runTest ("tests" </> "shouldwork" </> "Numbers") defBuild [] "BitInteger"   (["","testBench"],"testBench",True)
         , runTest ("tests" </> "shouldwork" </> "Numbers") defBuild [] "Bounds"       (["","testBench"],"testBench",True)
+        , runTest ("tests" </> "shouldwork" </> "Numbers") defBuild [] "DivideByZero" (["","testBench"],"testBench",True)
         , runTest ("tests" </> "shouldwork" </> "Numbers") defBuild ["-itests/shouldwork/Numbers", "-fconstraint-solver-iterations=15"] "ExpWithGhcCF"        (["","testBench"],"testBench",True)
         , runTest ("tests" </> "shouldwork" </> "Numbers") defBuild ["-itests/shouldwork/Numbers", "-fconstraint-solver-iterations=15"] "ExpWithClashCF"        (["","testBench"],"testBench",True)
         , outputTest ("tests" </> "shouldwork" </> "Numbers") defBuild ["-itests/shouldwork/Numbers"] ["-itests/shouldwork/Numbers"] "ExpWithClashCF"  "main"

